src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
+	/* the buffer allocated for the env string to append */
+	char  *env_buf_list[LEN_ENV_LIST] = {NULL};
+	char  *env_value                  = NULL;
+
+	if (envp == NULL) {
+		num_entry = 0;
+	} else if (envp[0] == NULL) {
+		num_entry = 0;
+	} else {
+		while (envp[num_entry]) {
+			/* scan the env in env_list[] to check whether they exist or not */
+			if (strncmp(envp[num_entry], STR_AND_SIZE_M1("LD_PRELOAD")) == 0) {
+				preload_included = true;
+				idx_preload      = num_entry;
+				num_entry_found++;
+				if (strstr(envp[num_entry], "libpil4dfs.so"))
+					pil4dfs_in_preload = true;
+			}
+			/* The list of env is not too long. We use a simple loop to lookup for
+			 * simplicity. This function is not performance critical.
+			 */
+			for (i = 0; i < LEN_ENV_LIST; i++) {
+				if (!env_found[i]) {
+					if (strncmp(envp[num_entry], STR_AND_SIZE_M1(env_list[i]))
+					    == 0) {
+						env_found[i] = true;
+						num_entry_found++;
+					}
+				}
+			}
+			num_entry++;
+		}
+	}
+
+	/* All required env are found and pil4dfs is in LD_PRELOAD. No need to create a new envp. */
+	if (num_entry_found == N_ENV_APPEND && pil4dfs_in_preload == true)
+		return (char **)envp;
+
+	/* the new envp holds the existing envs & the envs forced to append plus NULL at the end */
+	new_envp = malloc(sizeof(char *) * (num_entry + N_ENV_APPEND + 1));
+	if (new_envp == NULL) {
+		printf("Error: failed to allocate memory for new_envp. Use existing envp\n");
+		goto err_out0;
+	}
+
+	/* Copy all existing entries to the new envp[] */
+	for (i = 0; i < num_entry; i++) {
+		new_envp[i] = envp[i];
+	}
+
+	/* LD_PRELOAD is a special case. If LD_PRELOAD is found but libpil4dfs.so is not in
+	 * LD_PRELOAD env, then allocate a buffer for the concatenation of existing LD_PRELOAD
+	 * env and the full path of libpil4dfs.so.
+	 */
+	pil4df_path = query_pil4dfs_path();
+	len2        = strnlen(pil4df_path, PATH_MAX);
+	if (preload_included == true && pil4dfs_in_preload == false) {
+		/* need to replace the existing string */
+		len = strnlen(envp[idx_preload], MAX_ARG_STRLEN);
+		/* 2 extra bytes for ':' and '\0' */
+		len_total = len + len2 + 2;
+		if (len_total > MAX_ARG_STRLEN) {
+			printf("Error: env for LD_PRELOAD is too long.\n");
+			goto err_out1;
+		}
+		rc = asprintf(&new_preload_str, "%s:%s", envp[idx_preload], pil4df_path);
+		if (rc < 0) {
+			printf("Error: failed to allocate memory for LD_PRELOAD env.\n");
+			goto err_out1;
+		}
+		new_envp[idx_preload] = new_preload_str;
+	}
+	/* append LD_PRELOAD env */
+	if (!preload_included) {
+		rc = asprintf(&new_preload_str, "LD_PRELOAD=%s", pil4df_path);
+		if (rc < 0) {
+			printf("Error: failed to allocate memory for LD_PRELOAD env.\n");
+			goto err_out1;
+		}
+		new_envp[i] = new_preload_str;
+		i++;
+	}
+
+	for (j = 0; j < LEN_ENV_LIST; j++) {
+		/* env is not found. Need to be appended if present in current process. */
+		if (!env_found[j]) {
+			rc = d_agetenv_str(&env_value, env_list[j]);
```

Comment:
Does this work?  At this point if the environment has been cleared then this might fail?  In which case you'd need to sample the value in init() and keep it around to use it here.

Reply:
I think this should work. unsetenv() is not called in the case of scons, so env variables should be there.

Reply:
It seems likely that this will work for programs that call evecve without the env set but not for functions that call unsetenv and then execv, this is something we should test for and if necessary note.

Reply:
I believe applications call unsetenv() to clear only common system env or the env they set. I only observed "MAKEFLAGS" cleared by unsetenv() two times during daos building with scons.
We can intercept unsetenv(env) and ignore the operation if env is in our list and D_IL_ENFORCE_EXEC_ENV == 1.